cpp/include/cuml/experimental/fil/decision_forest.hpp

```diff
@@ -0,0 +1,466 @@
+   */
+  using categorical_storage_type = typename node_type::index_type;
+
+  /**
+   * Construct an empty decision forest
+   */
+  decision_forest() :
+    nodes_{},
+    root_node_indexes_{},
+    vector_output_{},
+    categorical_storage_{},
+    num_features_{},
+    num_outputs_{},
+    leaf_size_{},
+    has_categorical_nodes_{false},
+    row_postproc_{},
+    elem_postproc_{},
+    average_factor_{},
+    bias_{},
+    postproc_constant_{} {}
+
+  /**
+   * Construct a decision forest with the indicated data
+   *
+   * @param nodes A buffer containing all nodes within the forest
+   * @param root_node_indexes A buffer containing the index of the root node
+   * of every tree in the forest
+   * @param num_features The number of features per input sample for this model
+   * @param num_outputs The number of outputs per row from this model
+   * @param has_categorical_nodes Whether this forest contains any
+   * categorical nodes
+   * @param vector_output A buffer containing the output from all vector
+   * leaves for this model. Each leaf node will specify the offset within
+   * this buffer at which its vector output begins, and leaf_size will be
+   * used to determine how many subsequent entries from the buffer should be
+   * used to construct the vector output. A value of std::nullopt indicates
+   * that this is not a vector leaf model.
+   * @param categorical_storage For models with inputs on too many categories
+   * to be stored in the bits of an `index_t`, it may be necessary to store
+   * categorical information external to the node itself. This buffer
+   * contains the necessary storage for this information.
+   * @param leaf_size The number of output values per leaf (1 for non-vector
+   * leaves; >1 for vector leaves)
+   * @param row_postproc The post-processing operation to be applied to an
+   * entire row of the model output
+   * @param elem_postproc The per-element post-processing operation to be
+   * applied to the model output
+   * @param average_factor A factor which is used for output
+   * normalization
+   * @param bias The bias term that is applied to the output after
+   * normalization
+   * @param postproc_constant A constant used by some post-processing
+   * operations, including sigmoid, exponential, and
+   * logarithm_one_plus_exp
+   */
+  decision_forest(
+    raft_proto::buffer<node_type>&& nodes,
+    raft_proto::buffer<index_type>&& root_node_indexes,
+    index_type num_features,
+    index_type num_outputs=index_type{2},
+    bool has_categorical_nodes = false,
+    std::optional<raft_proto::buffer<io_type>>&& vector_output=std::nullopt,
+    std::optional<raft_proto::buffer<typename node_type::index_type>>&& categorical_storage=std::nullopt,
+    index_type leaf_size=index_type{1},
+    row_op row_postproc=row_op::disable,
+    element_op elem_postproc=element_op::disable,
+    io_type average_factor=io_type{1},
+    io_type bias=io_type{0},
```

Comment:
Note: we may need to support a vector `bias` in the future. See https://github.com/dmlc/treelite/issues/445

Reply:
Yep! Good to point out. If it's all right, I'd prefer to do that as a follow up once that change is in Treelite. We will need to address it soon, though.